### DISCUSSION TITLE
Further reducing boilerplate with... a protocol...?

### DISCUSSION THREAD

**Discussion by junebash** [pointfreeco/swift-composable-architecture#370] [Ideas]
Awesome work on the latest form-handling release! I always love removing boilerplate. :D

I realized that every time a binding is used with the new FormAction, `LocalAction.form` is passed in. I wonder if we can add a protocol the get rid of this boilerplate?

```
protocol FormActionHandling {
  associatedType State
  static func form(_ formAction: FormAction<State>) -> Self
}
```

Then, any action that handles forms can conform to this, and `case form(FormAction<State>)` would fulfill the protocol requirements. Then, a conditional extension on `ViewStore` could be made that removes the extra action argument `where Action: FormActionHandling`. This would remove the need to type `LocalAction.form` all over, but you would still be handling mutation explicitly in the reducer. You could also add an override on `Reducer.form()` where the `Action` conforms to `FormActionHandling`.

Thoughts? I'd be more than happy to draw up a PR for this if it seems like something that could be useful.

**Comment by siejkowski** [comment]
Great idea, just beware of a Swift bug causing the runtime crash when accessing the associated value of enum accesses through the protocol: https://bugs.swift.org/browse/SR-14071. I’ve been hit by this one when doing similar refactor.

**Comment by stephencelis** [comment]
This is definitely cool! I think we probably want to err on the side of not including protocols like these in the main library, but it definitely is handy enough that people might be interested in a micro-library or blog post. If you end up fleshing something out please do share with us!

Originally we sketched out a `binding` helper where the action was already a `FormAction` so that you could simply do:

```swift
text: viewStore.binding(\.displayName)
```

But in practice view stores do more than simple form actions, so it didn't seem practical enough to ship with TCA.

What we think this points to is the deficiency in Swift not having a "dynamic member lookup" for enum cases. If it did we could potentially leverage it here.

Edit: I think we will at the very least kick the tires a bit and see how such a protocol feels. We're usually cautious to use protocols in this way but it may prove fruitful.

**Comment by mbrandonw** [comment]
Just wanted to throw in that someone tweeted about this exact idea, and they also encountered the Swift bug that @siejkowski linked to :(

https://twitter.com/monadic_nomad/status/1356762575023792128

**Reply by siejkowski** [reply]
😞 I've also reported similar bug independently (https://bugs.swift.org/browse/SR-14041) and [Joe Groff's comment](https://bugs.swift.org/browse/SR-14041?focusedCommentId=60589&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-60589) explains that
> The crash is because of a calling convention mismatch between the protocol witness and the case constructor, which we can fix.

The last part brings some hope 😉

**Reply by atfelix** [reply]
This issue is fixed in Xcode 12.5.
